### ISSUE TITLE
[shared-state] Podman CLI crashes a container and corrupts podman state on the host, when executed inside (nested)

### ISSUE THREAD

**Issue by bogdando** [containers/libpod#1602]
**kind bug**

**Description**

When the podman CLI is used from a privileged container, it works only once, then the container crashes

**Steps to reproduce the issue:**

1. On a Centos 7.5 box with:

  * (Permissive SELinux),
  * / created as xfs,
  * running kernel 3.10.0-862.11.6.el7.x86_64,
  * installed podman 0.9.4-dev (podman-0.9.3.1-3.gita3c4ce6.el7.x86_64), libseccomp 2.3.1-3.el7

create a container:
```
sudo podman run -d --privileged -u root \
-v /etc/containers:/tmp/containers:ro \
-v /var/lib/containers:/var/lib/containers:rw \
-v /etc/yum.repos.d:/etc/yum.repos.d:ro \
-v /opt:/opt \
--name test centos:latest sleep 5000
```

2. Install required packages and configs into the container running (using the host versions/content)
```
sudo podman exec test yum -y install podman libseccomp
sudo podman exec test cp -af /tmp/containers /etc
```

3. Check the podman CLI within the container running, then repeat the command again
```
sudo podman exec test podman ps
sudo podman exec test podman ps
```

**Describe the results you received:**

First command worked, then the container crashed ``cannot exec into container that is not running``:
```
sudo podman inspect test | jq '.[] .State'
{
  "OciVersion": "1.0.1-dev",
  "Status": "configured",
  "Running": false,
  "Paused": false,
  "Restarting": false,
  "OOMKilled": false,
  "Dead": false,
  "Pid": 0,
  "ExitCode": 0,
  "Error": "",
  "StartedAt": "2018-10-08T08:23:21.031667631Z",
  "FinishedAt": "0001-01-01T00:00:00Z"
}
```
Then, all of the containers at the host start being shown as "Created" and nothing running (the host state for podman gets corrupted)

**Describe the results you expected:**

The container should keep running and serving podman CLI commands as if executed from the host.
There should be a way to not mess with the host state for podman.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:       0.9.4-dev
Go Version:    go1.10.2
OS/Arch:       linux/amd64
```

**Output of `podman info`:**

```
host:
  BuildahVersion: 1.4-dev
  Conmon:
    package: podman-0.9.3.1-3.gita3c4ce6.el7.x86_64
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.11.4, commit: 88d5276957861dea88b85f1d3f786612042cb4dd-dirty'
  Distribution:
    distribution: '"centos"'
    version: "7"
  MemFree: 384524288
  MemTotal: 8201936896
  OCIRuntime:
    package: runc-1.0.0-54.dev.git2abd837.el7.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.0'
  SwapFree: 4266651648
  SwapTotal: 8589930496
  arch: amd64
  cpus: 2
  hostname: undercloud.localdomain
  kernel: 3.10.0-862.11.6.el7.x86_64
  os: linux
  uptime: 91h 27m 11.39s (Approximately 3.79 days)
insecure registries:
  registries:
  - 192.168.24.1:8787
  - 192.168.24.3:8787
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  ContainerStore:
    number: 271
  GraphDriverName: overlay
  GraphOptions:
  - overlay.override_kernel_check=true
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
  ImageStore:
    number: 43
  RunRoot: /var/run/containers/storage
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by lachmanfrantisek** [comment]
I am playing with podman in podman for some time as well. (With some problems.)

I have a more general question: Are there some tips, what needs to be done to make nested podman work? It would be nice to have it somehow documented. It is hard to find a problem in my setup (e.g. missing option) from the low-level errors.

The only info I found is on #1475  and through FAH testing.

Thanks!

**Comment by cjeanner** [comment]
Hello,
Same kind of thing happening in here:
On a node with multiple deployed/running containers, running a simple "podman ps" from within a newly created container seems to crash the management.

Steps to reproduce on my own:

Running podman with the following command launches the container:
```Bash
run --rm --user root -ti --security-opt label=disable --cap-add SYS_ADMIN \
  -v /var/lib/containers:/var/lib/containers:rw \
  -v /etc/containers:/etc/containers:ro \
  -v /usr/bin:/usr/bin:ro \
  -v /dev/shm:/dev/shm:rw \
  -v /usr/libexec/podman/conmon:/usr/libexec/podman/conmon:ro \
  -v /usr/lib64/libseccomp.so.2:/usr/lib64/libseccomp.so.2:ro \
  -v /usr/lib64/libseccomp.so.2.3.1:/usr/lib64/libseccomp.so.2.3.1:ro \
  docker.io/tripleomaster/centos-binary-memcached:965941f1e62cef16967e7a7cd6d98263e52acb62_0989b280 /bin/bash
```
Running ```podman ps``` from the container doesn't show anything.
Running ```podman ps --all``` from the container shows the containers, as "created"
Running ```podman ps``` from the host doesn't show anything
Running ```podman ps --all``` from the host shows the containers, as "created"

When I want, from the host, manage the containers, it fails:
```podman start <id>``` returns:
```
DEBU[0000] [graphdriver] trying provided driver "overlay" 
DEBU[0000] overlay: override_kernelcheck=true           
DEBU[0000] overlay test mount with multiple lowers succeeded 
DEBU[0000] backingFs=xfs, projectQuotaSupported=false, useNativeDiff=true 
INFO[0000] Found CNI network podman (type=bridge) at /etc/cni/net.d/87-podman-bridge.conflist 
DEBU[0000] Initializing boltdb state at /var/lib/containers/storage/libpod/bolt_state.db 
DEBU[0000] Set libpod namespace to ""                   
DEBU[0000] mounted container "5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44" at "/var/lib/containers/storage/overlay/1fc2eb4933e142a2284c6b353151ee2487c5d2f4016ccffa6cb6e3f75031a6f0/merged" 
DEBU[0000] Created root filesystem for container 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 at /var/lib/containers/storage/overlay/1fc2eb4933e142a2284c6b353151ee2487c5d2f4016ccffa6cb6e3f75031a6f0/merged 
DEBU[0000] /etc/system-fips does not exist on host, not mounting FIPS mode secret 
WARN[0000] User mount overriding libpod mount at "/etc/hosts" 
WARN[0000] failed to parse language "en_US.utf8": language: tag is not well-formed 
DEBU[0000] reading hooks from /usr/share/containers/oci/hooks.d 
DEBU[0000] added hook /usr/share/containers/oci/hooks.d/oci-register-machine.json 
DEBU[0000] added hook /usr/share/containers/oci/hooks.d/oci-systemd-hook.json 
DEBU[0000] added hook /usr/share/containers/oci/hooks.d/oci-umount.json 
DEBU[0000] hook oci-register-machine.json did not match 
DEBU[0000] hook oci-systemd-hook.json did not match     
DEBU[0000] hook oci-umount.json matched; adding to stages [prestart] 
DEBU[0000] reading hooks from /etc/containers/oci/hooks.d 
WARN[0000] failed to load hooks: {}%!(EXTRA *os.PathError=open /etc/containers/oci/hooks.d: no such file or directory) 
DEBU[0000] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] exporting opaque data as blob "sha256:22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] exporting opaque data as blob "sha256:22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] parsed reference into "[overlay@/var/lib/containers/storage+/var/run/containers/storage:overlay.override_kernel_check=true]@22cd549d26e2b1b17d99deb0f8ead6af5f5b0fb84e01b233a9294e99ae21d719" 
DEBU[0000] Setting CGroups for container 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 to machine.slice:libpod:5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 
DEBU[0000] Created OCI spec for container 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 at /var/lib/containers/storage/overlay-containers/5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44/userdata/config.json 
DEBU[0000] /usr/libexec/podman/conmon messages will be logged to syslog 
DEBU[0000] running conmon: /usr/libexec/podman/conmon    args=[-s -c 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 -u 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 -r /usr/bin/runc -b /var/lib/containers/storage/overlay-containers/5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44/userdata -p /var/run/containers/storage/overlay-containers/5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44/userdata/pidfile -l /var/lib/containers/storage/overlay-containers/5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44/userdata/ctr.log --exit-dir /var/run/libpod/exits --exit-command /usr/bin/podman --exit-command-arg --root --exit-command-arg /var/lib/containers/storage --exit-command-arg --runroot --exit-command-arg /var/run/containers/storage --exit-command-arg --log-level --exit-command-arg error --exit-command-arg --cgroup-manager --exit-command-arg systemd --exit-command-arg --tmpdir --exit-command-arg /var/run/libpod --exit-command-arg --storage-driver --exit-command-arg overlay --exit-command-arg container --exit-command-arg cleanup --exit-command-arg 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 --socket-dir-path /var/run/libpod/socket --log-level debug --syslog]
INFO[0000] Running conmon under slice machine.slice and unitName libpod-conmon-5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44.scope 
DEBU[0000] Received container pid: 88617                
DEBU[0000] Created container 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 in OCI runtime 
DEBU[0000] Started container 5298aaf3e2cf999c432efd49a0eccacbbbfb1ed753fe96a483040752105a8d44 
5298aaf3e2cf
```

I have to run ```podman stop <id>``` in order to get it in "stopped" state, then only I can start it again with ```podman start <id>```.
And the ```podman stop <id>``` returns an error stating:
```
failed to stop container <id>: can only stop created, running, or stopped containers: container state improper
```

So apparently there's something breaking state supervision at some point.

In my case, the podman version is: podman version 0.9.4-dev

Cheers,

C.

**Comment by rhatdan** [comment]
@mheon This looks like a fun one for you.  :^(

**Comment by giuseppe** [comment]
if you are going to share the storage with the host, then you will need to ensure the podman in a container has full access to the status of the other containers.  Things you'll need to add:

- share more namespaces with the host (at least PID and NETWORK, probably IPC as well)
- bind mount `/sys/fs/cgroup`
- bind mount `/var/run/containers`
- `/var/lib/containers` must be bind mounted with the `rshared` option.

This command worked for me on Fedora:

```
# podman run -d --net=host --uts=host  --ipc=host --pid=host  --privileged  -v /sys/fs/cgroup:/sys/fs/cgroup:rshared -v /etc/containers:/etc/containers:ro -v /var/lib/containers:/var/lib/containers:rw,rshared -v /var/run/libpod:/var/run/libpod:rshared,rw -v /run:/run:rshared,rw  --name test fedora sleep 5000
```

then I can use `podman ps` from the container

**Comment by rhatdan** [comment]
It is weird that the container should get killed, but I guess we are rewriting the podman database to say that their are no containers, since it is looking at /run inside of the container and see no containers are running.

**Comment by mheon** [comment]
As Giuseppe said, this is caused by namespaces. We need access to host net
and PID to get container status and clean up when they exit.

(This should probably be documented in a blog somewhere)

**Comment by cjeanner** [comment]
Oh, thanks for the headsup regarding the needed mounts. Maybe podman could error in some nice fashion instead of overriding the state?

**Comment by rhatdan** [comment]
Not sure how it would recognize this. The problem here is you had partial leakage of the host into the container.  From the podman point of view inside of the container everything looked good except that the containers listed in the database did not meet the reality in its environment,  It fixes its database. and since the database is shared with the host, it screws up the host.

If we start making assumptions about how it should work, then we will most likely break the people who want to use podman inside of a container.

I think this is a documentation issue.  If you want to use podman to manage the hosts podman environment here is how you need to set it up...

**Comment by cjeanner** [comment]
Right. Go for a doc issue :). @bogdando OK for you as well?

**Comment by bogdando** [comment]
folks, given that podman overwrites and clobbers the host state for all of the containers, I'd say that is not a documentation bug, unless we going to put there "not supported, please user varlink instead".

Otherwise, I can foresee a situation when simultaneous access to those rshared bind-mounted state paths, makes all the data corrupted.

**Comment by mheon** [comment]
Agree that this is a docs issue - we need a proper description of what needs to be shared into a container to make podman-in-podman with a shared DB functional

**Comment by cjeanner** [comment]
small addition: instead of sharing the /run (aka /var/run), we can share precisely the following bind:
```
-v /run/libpod:/run/libpod:rshared,rw
-v /run/containers:/run/containers:rshared,rw
-v /run/netns:/run/netns:rshared,rw
-v /run/runc:/run/runc:rshared,rw
-v /run/runc-ctrs:/run/runc-ctrs:rshared,rw
```

But still... this represent a risk of corruption at some point, and that's not really "nice" :(.

**Comment by giuseppe** [comment]
Another thing to consider here is that we could potentially have different versions of podman accessing the same storage, and that might be a problem.

For now that we haven't a read-only mode for podman, instead of cherry-picking the right mounts, I think it will be safer to share the host with the container (-v /:/host:rw,rslave) then from the container
chroot to /host and use podman from there.
Once podman won't require write access to the storage we can switch the bind mount to 'ro'.

**Comment by bogdando** [comment]
^^ A good option to at least to be protected from adding more paths in future that will need to be propagated inside the container as well

**Comment by giuseppe** [comment]
or even "safer", `# nsenter --all -t 1 podman ...` so you will really execute podman in the same namespaces as on the host and there should be no difference at all (with the chroot aproach private mounts will not be propagated back)

**Comment by rhatdan** [comment]
@mheon What do you want to do with this issue?

**Comment by mheon** [comment]
@rhatdan I think we need some documentation written on this

**Comment by rhatdan** [comment]
@mheon Did you ever write anything up?

**Comment by mheon** [comment]
Negative, let me get something in the Troubleshooting guide
